## 제주도 식당정보의 도로명주소를 통해 네이버 지오코딩으로 위도경도 받아오기
- 제주_식당정보.csv를 사용

---
### 불러오기

In [10]:
import pandas as pd

In [11]:
jejuData = pd.read_csv("Data/제주_식당정보.csv", encoding="cp949")

--- 
### 데이터 확인해보기

In [12]:
jejuData.head()

사업장명 업종구분대분류   업종구분소분류     인허가일자  인허가취소일자  영업상태명 상세영업상태명  폐업일자  \
0      탑동에이치지   일반음식점        기타  20181129      NaN  영업/정상      영업   NaN   
1    에릭스에스프레소   일반음식점        기타  20181102      NaN  영업/정상      영업   NaN   
2    일품순두부한림점   일반음식점        한식  20190107      NaN  영업/정상      영업   NaN   
3  김복남맥주제주아라점   일반음식점  식육(숯불구이)  20190107      NaN  영업/정상      영업   NaN   
4       봉플라봉뱅   일반음식점        기타  20190107      NaN  영업/정상      영업   NaN   

   휴업시작일자  휴업종료일자  재개업일자   소재지면적                        소재지전체주소  \
0     NaN     NaN    NaN   80.14        제주특별자치도 제주시 삼도이동 1120-2   
1     NaN     NaN    NaN   27.00   제주특별자치도 제주시 구좌읍 세화리 1397-6번지   
2     NaN     NaN    NaN  152.00   제주특별자치도 제주시 한림읍 대림리 1845-2번지   
3     NaN     NaN    NaN   82.00          제주특별자치도 제주시 아라일동 6139   
4     NaN     NaN    NaN   58.94  제주특별자치도 제주시 연동 312-57번지 정인하우스   

                               도로명전체주소  도로명우편번호     데이터갱신일자  
0      제주특별자치도 제주시 무근성안길 16, 1층 (삼도이동)  63166.0  2023-11-07  
1           제주특별자치도 제주시 구좌읍 구좌로 77, 1층  63361.0  2023-11-07  
2         제주특별자치도 제주시 한림읍 한림상로 237, 1층  63030.0  2023-11-07  
3   제주특별자치도 제주시 중앙로 581, 에이동 1층 (아라일동)  63248.0  2023-11-07  
4  제주특별자치도 제주시 문송1길 6-1, 정인하우스 1층 (연동)  63122.0  2023-11-07

In [13]:
jejuData.count(axis=1)

0        11
1        11
2        11
3        11
4        11
         ..
47562    12
47563    12
47564    12
47565    12
47566    12
Length: 47567, dtype: int64

> 식당 데이터는 총 47,567개

### 결측치 체크하기

In [14]:
jejuData['폐업일자'].isna().sum()

20876

> 47,567개 중에서 폐업하지 않은 식당 갯수는 20,876개

In [21]:
jejuData[jejuData['도로명전체주소'].isna()]

사업장명 업종구분대분류 업종구분소분류     인허가일자  인허가취소일자  영업상태명 상세영업상태명  폐업일자  \
8              레게펍   일반음식점      기타  20181105      NaN  영업/정상      영업   NaN   
522            타베니   일반음식점      한식  20190614      NaN  영업/정상      영업   NaN   
589      서우봉통돼지바베큐   일반음식점      한식  20230626      NaN  영업/정상      영업   NaN   
590           올레포차   일반음식점      분식  20230626      NaN  영업/정상      영업   NaN   
592          머물소낭밭   일반음식점      한식  20230626      NaN  영업/정상      영업   NaN   
...            ...     ...     ...       ...      ...    ...     ...   ...   
45724         하늬바람   휴게음식점    푸드트럭  20210408      NaN  영업/정상      영업   NaN   
45846      진모살푸드트럭   휴게음식점    푸드트럭  20201224      NaN  영업/정상      영업   NaN   
45857       색달푸드트럭   휴게음식점    푸드트럭  20201201      NaN  영업/정상      영업   NaN   
45912  제주밭담샾어멍아방카페   휴게음식점     커피숍  20230921      NaN  영업/정상      영업   NaN   
45916      돌고래푸드트럭   휴게음식점    푸드트럭  20230927      NaN  영업/정상      영업   NaN   

       휴업시작일자  휴업종료일자  재개업일자   소재지면적  \
8         NaN     NaN    NaN   30.20   
522       NaN     NaN    NaN   57.36   
589       NaN     NaN    NaN  163.00   
590       NaN     NaN    NaN  100.00   
592       NaN     NaN    NaN   60.00   
...       ...     ...    ...     ...   
45724     NaN     NaN    NaN    9.48   
45846     NaN     NaN    NaN   12.00   
45857     NaN     NaN    NaN   12.00   
45912     NaN     NaN    NaN   90.00   
45916     NaN     NaN    NaN    9.25   

                                           소재지전체주소 도로명전체주소  도로명우편번호  \
8                         제주특별자치도 제주시 한림읍 귀덕리 3019     NaN      NaN   
522                제주특별자치도 제주시 애월읍 납읍리 2340-2번지 다동     NaN      NaN   
589                제주특별자치도 제주시 조천읍 함덕리 산 1004 외1필지     NaN      NaN   
590                  제주특별자치도 제주시 조천읍 함덕리 1004 외1필지     NaN      NaN   
592                       제주특별자치도 제주시 애월읍 곽지리 1565     NaN      NaN   
...                                            ...     ...      ...   
45724                 제주특별자치도 서귀포시 남원읍 위미리 산 102-5     NaN      NaN   
45846                      제주특별자치도 서귀포시 색달동 2950-3     NaN      NaN   
45857                      제주특별자치도 서귀포시 색달동 2950-3     NaN      NaN   
45912                   제주특별자치도 서귀포시 성산읍 신풍리 456-8     NaN      NaN   
45916  제주특별자치도 서귀포시 법환동 914 제주월드컵경기장 광장 내(별도 지정구역)     NaN      NaN   

          데이터갱신일자  
8      2023-11-07  
522    2023-11-07  
589    2023-11-07  
590    2023-11-07  
592    2023-11-07  
...           ...  
45724  2023-11-07  
45846  2023-11-07  
45857  2023-11-07  
45912  2023-11-07  
45916  2023-11-07  

[220 rows x 16 columns]

In [22]:
jejuData['소재지전체주소'].isna().sum()

76

In [27]:
jejuData['사업장명'].isna().sum()

0

---
### DataPreprocessing

In [18]:
# 폐업일자가 없는 데이터만 쓰기
jejuData = jejuData[jejuData['폐업일자'].isna()]

> 도로명 주소가 na값인 것들이 220개여서 그냥 제외하기

In [ ]:
# 도로명주소가 있는 데이터만 쓰기
jejuData = jejuData[~jejuData['도로명전체주소'].isna()]

In [26]:
jejuData.count(axis=1)

0        11
1        11
2        11
3        11
4        11
         ..
45910    11
45911    11
45913    11
45914    11
45915    11
Length: 20656, dtype: int64

> 쓸 데이터는 최종적으로 20,656개임

---
### 네이버 지오코딩으로 도로명 전체주소를 써서 위도경도 뽑기

In [31]:
import urllib.request # network 요청 보내기
import json
import time
import random

# 네이버 API 키 정보들
client_id = "zh4c9yakmm"
client_secret = "Nh6EIvxVKC6LhvKCpojoHOEfdTCArsL1ET3tqa70"

In [29]:
# 위도경도를 뽑아올 함수 정의
def addressToLatLng(address):
    url = "https://naveropenapi.apigw.ntruss.com/map-geocode/v2/geocode?query=" + urllib.parse.quote(address)
    request = urllib.request.Request(url)
    request.add_header("X-NCP-APIGW-API-KEY-ID", client_id)
    request.add_header("X-NCP-APIGW-API-KEY", client_secret)
    response = urllib.request.urlopen(request)
    rescode = response.getcode()
    lat_lng = None
    if rescode == 200:
        response_body = response.read()
        entext = response_body.decode('utf-8')
        data = json.loads(entext)
        try:
            lat_lng = {
                "lat": data['addresses'][0]['y'],
                "lng": data['addresses'][0]['x']
            }
        except:
            # 주소에서 가져올 수 없을 때 공백 추가
            lat_lng = {"lat": "", "lng": ""}
    else:
        print("Error Code:" + rescode)
    return lat_lng
# 데이터프레임 초기화
df = pd.DataFrame(columns=['latitude', 'longitude'])

In [34]:
# Test
# addressToLatLng("제주특별자치도 제주시 중앙로 581, 에이동 1층 (아라일동)")
count = 1
# 주소별 위도와 경도를 추출하여 데이터프레임에 추가
for address in jejuData['도로명전체주소']:
    time.sleep(random.uniform(0.5, 1.0)) 
    lat_lng = addressToLatLng(address)
    if lat_lng:
        df = pd.concat([df, pd.DataFrame([lat_lng])], ignore_index=True)        
        print(count)
        count += 1

print("완료")

1
2
3
4
5
6
7
8
9
10
11
12
13


KeyboardInterrupt: 